src/tagged-type.ts

```diff
@@ -0,0 +1,117 @@
+export type Module = string;
+export type DocumentReference = string;
+
+/**
+ * TaggedType provides the encoding/decoding of the Fauna Tagged Type formatting
+ */
+export class TaggedTypeFormat {
+  /**
+   * Encode the Object to the Tagged Type format for Fauna
+   *
+   * @param obj - Object that will be encoded
+   * @returns Map of result
+   */
+  static encode(obj: any): any {
+    return new TaggedTypeEncoded(obj).result;
+  }
+
+  /**
+   * Decode the JSON string result from Fauna to remove Tagged Type formatting.
+   *
+   * @param input - JSON string result from Fauna
+   * @returns object of result of FQL query
+   */
+  static decode(input: string): any {
+    return JSON.parse(input, (_, value: any) => {
+      if (value["@mod"]) {
+        return value["@mod"] as Module;
+      } else if (value["@doc"]) {
+        return value["@doc"] as DocumentReference;
+      } else if (value["@int"]) {
+        return Number(value["@int"]);
+      } else if (value["@long"]) {
+        return Number(value["@long"]);
+      } else if (value["@decimal"]) {
+        return Number(value["@decimal"]);
+      } else if (value["@double"]) {
+        return Number(value["@double"]);
+      } else if (value["@date"]) {
+        return new Date(value["@date"]);
+      } else if (value["@time"]) {
+        return new Date(value["@time"]);
+      } else if (value["@object"]) {
+        return value["@object"];
+      }
+
+      return value;
+    });
+  }
+}
+
+class TaggedTypeEncoded {
+  readonly result: any;
+
+  readonly #encodeMap = {
+    number: (value: number): { [key: string]: number } => {
+      if (`${value}`.indexOf(".") > 0) {
+        return { "@decimal": value };
+      } else {
+        return { "@int": value };
+      }
+    },
+    string: (value: string): string => {
+      return value;
+    },
+    object: (input: any): { [key: string]: string } => {
+      const _out: { [key: string]: any } = {};
+      for (const k in input) {
+        if (k.startsWith("@")) {
+          _out[k] = input[k];
+        } else {
+          _out[k] = TaggedTypeFormat.encode(input[k]);
+        }
+      }
+      return _out;
+    },
```

Comment:
This I think needs some work.

- the result should be wrapped in an `"@object"` field. 
- fields provided from the user, starting with `@` only matter if the field is an object and contains exactly one field starting with `@`. If there are multiple fields starting with `@` we can interpret the object as having user-defined fields that happen to start with `@`

Reply:
> We can't assume that fields starting with @ are Fauna tagged fields.

We're not. Per the [spec](https://github.com/fauna/core/blob/master/ext/api/src/main/resources/openapi.yml#L442-L445):
> If an FQL object value contains properties starting with `@`, it will be emitted on the wire in an `@object` tag.

It one of the fields of their object has an `@` prefix, we just pass it through as an object rather than applying any tagging to it.

Reply:
For example `{ "@foo": new Date("2022-12-02") }` should encode to 

```JSON
{
  "@object": {
    "@foo": {
      "@date": "2022-12-02"
    }
  }
}
```

but will currently encode to 

```JSON
{
  "@foo": "2022-12-01T05:00:00.000Z"
}
```

Reply:
So, the `@object` tag is not strictly necessary when there are no other "@" fields, but it is necessary when there are tagged types. It might be easier if the driver just always wraps in `@object`

we can provide an `foo` object field without the `@object` wrapper if there are no user-defined "@" fields

![image](https://user-images.githubusercontent.com/7998974/220440268-d0ca1ccc-0a47-4332-aeba-5b43ccb57eea.png)

if we rename the field to `@foo` we get an error.

![image](https://user-images.githubusercontent.com/7998974/220439805-83d057e4-0949-4175-8f0a-21d289a3622b.png)

but we can fix it by wrapping in `@object`

![image](https://user-images.githubusercontent.com/7998974/220440645-df1bb009-de6f-44d1-bf9f-57427b67303b.png)

Reply:
I see what you're saying now, thanks for the screenshots w/the curl examples. @pnwpedro given this we should update the Python driver encoding too.

Reply:
Does [this example](https://github.com/fauna/fauna-python/blob/dev/tests/fixtures/tag_conflicts.py#L56-L80) in the python driver mean we're covered?

Reply:
@pnwpedro  I agree. Python and JS should be in sync an both covering , considering the test you linked to and [the encoding for dicts](https://github.com/fauna/fauna-python/blob/0dfee551c312f0689d08e9bacc42349e6c75c00b/fauna/encode.py#L68-L72) lines up with JS, too.